### ISSUE TITLE
0.6.3: Qwen ANE prefill corrupts context under agentic load (jumbled prompt content, empty init-only completions); clean with ANE off, Lightning MTP on

### ISSUE THREAD

**Issue by daveycodez** [jundot/omlx#3233]
## Environment

- oMLX **0.6.3** (final, updated same day from an 0.6.3rc)
- Mac Studio M3 Ultra, 512 GB
- Model: `Jundot/Qwen3.8-27B-oQ4e-fp16-mtp` (discovered as `type: vlm`, served by VLMBatchedEngine)
- Per-model settings: Lightning MTP on (`mtp_enabled: true`, adaptive depth), **Qwen ANE prefill on** (tuner-calibrated: MLP 45%, GDN 37.5%, dual-ANE, CPU share), `qwen3-r-code` profile (temp 0.6, top_p 0.95, thinking on), TurboQuant/DFlash/SpecPrefill off
- Workload: agentic coding harness against the Anthropic-compatible endpoint — multi-step tool-calling turns, 10–30K-token prefills per step, prompt-cache reuse 50–90%, session contexts 90K–350K

## Summary

With **ANE prefill + Lightning MTP both enabled**, agentic sessions reliably derail within 5–14 steps (two distinct symptoms below; reproduced on both warm and cold prompt cache, including immediately after deleting the SSD cache).

**Isolation so far: ANE prefill OFF, Lightning MTP still ON → a comparable session (8 steps, ~209K cumulative input, 75% cache hit) ran to completion with a correct long-form final answer at ~41 tok/s.** So the corruption tracks the `qwen35_ane_prefill` path, not MTP itself. (n=1 clean run so far; will add more runs as I have them.)

## Symptom 1: model receives corrupted prompt content mid-session

Deep into a session (step 14, ~350K cumulative input, ~85% cache hit), the model began reasoning about tool output it was never sent: its visible reasoning quoted short repeated fragments (a numeric token `262144` = 2^18, plus one filename repeated several times with stray line-number-like values) that do not appear in any tool result recorded in the harness transcript — the fragments look like re-spliced tokens from much earlier context. The model itself remarked the output "looks jumbled" and attempted to re-read the file; that recovery tool call was emitted with broken framing (literal `</parameter> </function> </tool_call>` closers leaked into the text channel, no parseable call), then `stopReason: stop`. All 13 prior steps had perfectly-formed tool calls.

## Symptom 2: empty completion (EOS as first sampled token)

In a later session, a step prefilled normally (~13 s) and returned an empty completion (empty reasoning + text, immediate `stop`). Server log for that request:

```
MTP[26] finish=stop tokens=1 cycles=0 tok/cycle=0.00 accept=0/0 (n/a) emits[init=1,draft=0,bonus=0,verify=0] timing[backbone=0.0ms mtp=0.0ms sample=0.0ms cache=0.0ms]
Anthropic message: model=Qwen3.8-27B-oQ4e-fp16-mtp, 0 tokens in 13.28s (0.0 tok/s)
```

I initially read this as an MTP batch-generator bug, but given the isolation result it's more consistent with a downstream symptom: if the ANE-split prefill writes corrupted KV state, EOS as the first sampled token is a plausible model response, and the MTP loop then correctly logs an init-only sequence. A similar one-token anomaly (`finish=late-join-handoff tokens=1`) and repeated `MTP singleton recovery: multirow marker cleared (compact cache)` lines appeared in the same sessions.

## Notes

- Lightning MTP alone looks healthy and fast here: acceptance 60–98% depending on content, ~46 tok/s vs ~31 baseline short-context, and the ANE-off session above completed cleanly with MTP still enabled.
- ANE prefill also looked correct in isolation on cold bulk prefill (dashboard ~503 tok/s on large uncached chunks, close to the tuner's 565 tok/s prediction) — the failures only showed up under the agentic traffic shape (many sequential 4K+ chunks, partial tails, heavy cache-block reuse, MTP active in the same session).
- Harness session transcripts (JSONL event streams) for both failing runs and the clean ANE-off run are saved; happy to share privately (they contain project source, so not attaching publicly).

Happy to run instrumented builds or additional repros — this machine reproduces the failures quickly, and ANE prefill's +26% PP is very much worth having back alongside MTP.

**Comment by daveycodez** [comment]
Follow-up isolation runs, as promised: with ANE prefill OFF and Lightning MTP ON, three further agentic sessions (137K–209K cumulative input each, 72–84% prompt-cache hit, 6–8 tool-calling steps) all ran to completion with correct long-form final answers — two on the oQ4e conversion (~41 tok/s) and one on oQ8e (~32 tok/s). No recurrence of either symptom (no corrupted prompt content, no init-only empty completions). That's now 4/4 clean with ANE off vs. reliable derailment within 5–14 steps with ANE on, same model, same workload. Confident the corruption tracks the qwen35_ane_prefill path.

**Comment by daveycodez** [comment]
Substantial update after the fixes in PR #3246: the corruption reported here may not be ANE prefill's fault primarily. Before those fixes, ANE-on agentic sessions reliably derailed within 5-14 steps. After applying the four continuous-batching/MTP fixes from #3246, an A/B with a purpose-built repro (multi-turn agentic traffic: 13-16K cached contexts, odd-length partial-chunk tails, a concurrent sidecar request joining every turn, needle-retrieval verification each turn, Lightning MTP on) gives: ANE ON 32 turns / 1 failure (a single silent empty completion, no log entries), ANE OFF 20 turns / 0 failures. The catastrophic failure mode is gone with ANE on; the remaining signal is statistically indistinguishable from zero. Suggests the corruption attributed to ANE prefill was mostly the batch-join/MTP family fixed in #3246, with ANE prefill acting as a timing catalyst. Recommend re-evaluating this issue after merging #3246; the one residual silent-empty in 32 turns may deserve its own watch but is not clearly ANE-linked.

**Comment by hesara** [comment]
I can reproduce what appears to be the same issue on current `main`
(`a20ec09d`, oMLX 0.6.3), using Qwen3.8-27B-oQ4e-mtp on an M3 Max.

With MLP ANE prefill and prefix-cache reuse enabled, a later request can begin
generating text copied from an earlier assistant tool call and eventually emit
literal closing tool markup. MTP is not required. The failure is deterministic
with `temperature=0` and visible in the first 16 output tokens, so long decoding
and tool-call parsing are not prerequisites.

In my tested controls, the issue did not reproduce with GPU prefill or with the
prefix cache disabled.

Hash-only instrumentation at the failing 24,576-token boundary found that all
128 cache arrays—32 KV and 96 recurrent-state arrays—were byte-identical between
the live prefill cache and the reconstructed six-block cache. Two independent
ANE prefills of the same prefix also produced identical cache arrays and logits.
This suggests that neither ANE nondeterminism nor corruption during cache
storage/reconstruction explains the failure.

A synthetic single-message prompt with the same six cached blocks exactly
matched a fresh-prefill control, so six-block reuse or prefix length alone is
not sufficient. A reduced realistic case still reproduces after replacing the
original system/user instructions and retaining only `write` and `bash` tools.
Its cache-hit response immediately starts inside the earlier large `write`
argument.

I am continuing later tonight or tomorrow to reduce that tool-history case
and investigate how the correctly reconstructed cache is consumed during
suffix prefill.

**Comment by beaglemoo** [comment]
Follow-up with a localization that may shorten the reduction. Same rig as my earlier comment (M5 Pro, NAX, `c0503a9e` = the 0.6.3 ANE code, `mtp_enabled: false`, temperature 0, dual-ANE, cap active), same flow: prime the cache with a short document sharing the first ~3.2K tokens, then send the 32K document, so every run restores `reused 2048` and re-prefills 29,886 tokens. Two fresh processes per row, ANE state verified armed and unshed before every arm, no tiled-sdpa fallback in any of these logs.

| ANE offload config | restore run A vs B | restore vs fresh full prefill (same config) |
|---|---:|---:|
| GPU-only | 5/5 | 5/5 |
| MLP only (64 layers, `gdn: false`) | **5/5** | 0/5 (deterministic offset) |
| GDN + 1 MLP layer (`max_layers: 1`, 48 GDN) | **5/5** | **5/5** |
| MLP 64 + GDN 48 (the normal config) | 0/5 and 1/5 | 0/5 |

So neither bank is non-deterministic by itself. The MLP offload after a restore is byte-stable but differs from the full-prefill result (GPU-only and GDN-offload do not differ), and only with both banks active in the same forward does the result become non-reproducible run to run. A prior unrelated 6K-token ANE prefill does not change a subsequent full prefill (5/5), so there is no state carried inside the programs; whatever it is lives in how the restored-prefix suffix is scheduled across the two banks.

Two negatives that may save someone time:

- `dual_ane: false` does not help (single-instance pair: 2/5 agreement, 0/5 vs full).
- Draining the merge command buffer before returning from each hybrid dispatch — commit plus host wait, so a program's output surface can never be overwritten by the next tile before its merge runs — did not change the outcome either (1/5, 0/5). I had read that as a candidate race from the single output surface per program; on this evidence it is not the mechanism.

One reading of the table against hesara's result: their failure is deterministic and on the MLP offload, and the MLP-only row here is exactly that shape — byte-stable, but not what full prefill produces — while the run-to-run variance needs both banks. So there may be two things: a deterministic MLP-offload-after-restore divergence (theirs, catastrophic on M3 Max; a phrasing-level offset here), and a scheduling interaction between the banks on top of it. `qwen35_ane_prefill_gdn: false` removes the second here and is worth a try for the agent-session reports, but on this evidence it should not be expected to remove the first. Full runner and per-arm JSON available on request.

**Comment by hesara** [comment]
What beaglemoo saw is consistent with my deterministic case. I've localized my reproduction further to the padded MLP tail used during scheduler prefill after a prefix-cache restore.

On an M3 Max with Qwen3.8-27B-oQ4e-mtp and current `main` (`a20ec09d`), my smallest server-level case has MTP disabled, `temperature=0`, and only 8 generated tokens. A 5,934-token probe restores one 4,096-token block and leaves 1,838 tokens. The scheduler reserves the final token and prefills a 1,837-token tail; because this exceeds the calibrated 1,593-token threshold, it is zero-padded to the fixed 2,048-token ANE shape.

With tail padding enabled, the response incorrectly begins by continuing earlier assistant/tool content:

`Let me write the update function and the`

Disabling only ANE tail padding produces the expected continuation:

`The syntax is fine. Next, I`

I might get back to digging deeper tomorrow.

**Comment by beaglemoo** [comment]
hesara's tail-padding localization reproduces on NAX, with one extra discriminator.

M5 Pro 64 GB, `c0503a9e`, `Qwen3.8-27B-oQ4e-mtp`, MLP offload only (64 layers, `gdn: false`, dual-ANE), `mtp_enabled: false`, temperature 0, `OMLX_SDPA256_TILED=0` pinned, fresh process and empty cache per arm. This rig normally runs `tail_padding_min_tokens: 0`; setting it to hesara's calibrated 1593 is the only change between the arms.

| Flow | tail padding 1593 | tail padding 0 | GPU-only |
|---|---|---|---|
| Restore 4096 tokens, re-prefill a ~1,880-token tail (5 probes) | **0 tokens, `finish_reason=stop`, all 5 probes, byte-identical on a second fresh process** | 48 tokens each, correct | 48 tokens each, correct (4/5 byte-identical to padding-0) |
| Fresh 5,979-token prompt: 2 full tiles + the same ~1,880-token padded tail, no restore | 48 tokens, correct | 48 tokens, correct | — |
| Same document served again from the cache (restore + padded tail) | **0 tokens, `stop`, 4/4** | 48 each | — |
| Fresh ~1,950-token prompt (padded whole, no restore) | 48 tokens, correct, 5/5 | 48 each | — |

So the padded ANE MLP tile is fine on a fresh prefill and wrong only when it is the first chunk after a prefix-cache restore — then it yields an immediate EOS here rather than hesara's "continue earlier assistant content" on M3 Max, which is the #3117 symptom exactly. The MLP projection is position-independent, so whatever differs must be in what that first post-restore dispatch receives or how its padded rows are merged, not in the arithmetic; the same rows through the GPU tail path are correct.

Two consequences worth stating: `tail_padding_min_tokens: 0` (tail on the GPU) is a clean workaround with a small prefill cost on tails only, and the 0.6.3 tuner's calibrated tail padding is what puts users on this path by default. Runner and per-arm JSON available on request.

**Comment by hesara** [comment]
I think I found the root cause. The trigger is not in the ANE arithmetic or the cache round trip but a lazy cross-stream array that only exists on the restored-cache path.

**Mechanism.** For text-only requests continuing from a restored prefix cache (`cached_tokens > 0`), the scheduler seeds the model's mRoPE state with `lm._rope_deltas = mx.zeros((1, 1))` (`_seed_text_only_mrope_delta_for_cached_prefill`, scheduler.py). That array is created lazily and outside the `with mx.stream(engine_stream)` context that runs the prefill chunks. With a nonzero cache offset, the qwen3_5 language model folds this seed into `delta = mx.array(cache_offset + rope_deltas)`, i.e. into every RoPE position of the chunk forward — so the entire chunk graph carries a cross-stream dependency. The padded ANE MLP path then calls `waitUntilCompleted()` mid-eval on the producer command buffer that carries that dependency (`qwen35_ane.mm`, input-pack step). Depending on machine and timing this either deadlocks (a `kIOGPUCommandBufferCallbackErrorTimeout` in my direct scheduler harness) or the pack reads input buffers that are not yet written — producing deterministic, machine-specific garbage-in: old-context continuation on my M3 Max, immediate EOS on beaglemoo's M5 Pro.

My direct/hand-emulated restored-cache padded prefills were always byte-exact as they never seeded `_rope_deltas`, which is why cache-serialization and ANE-determinism tests kept coming back clean. The corruption seems to be byte-stable per machine but differs across machines: a reproducible stale-buffer read, not nondeterminism.

**Confirmation.** A one-line change - materializing the seed with `mx.eval` at creation, nothing else, tail padding still enabled at the calibrated 1,593 threshold - fixes both of my reproducers on M3 Max at `a20ec09d`:

- The direct `Scheduler._do_external_prefill()` harness (which previously hit the Metal timeout) completes and selects the expected first token `The`.
- The full server replay of my minimized 5,934-token pair returns the correct continuation `The syntax is fine. Next, I` instead of `Let me write the update function and the`, deterministically at `temperature=0`.

**Fix.** Create/evaluate the `_rope_deltas` seed on the engine stream (one line in the scheduler). Calibrated tail padding can stay enabled; no ANE fallback is needed. Until fixed, one can mitigate this with `tail_padding_min_tokens: 0`. The 0.6.3 tuner's calibrated padding is what puts users on the failing path by default.

@beaglemoo - could you confirm that adding `mx.eval(lm._rope_deltas)` right after the seed assignment in `_seed_text_only_mrope_delta_for_cached_prefill` (scheduler.py) turns your restore + padded-tail arm from immediate-EOS to the correct 48 tokens, with your normal calibrated threshold?

**Comment by beaglemoo** [comment]
Confirmed on the M5 Pro, and it explains more than the padded tail.

Applied exactly `mx.eval(lm._rope_deltas)` after the seed in `_seed_text_only_mrope_delta_for_cached_prefill`, on `c0503a9e`, nothing else changed.

**Your ask — restore + padded tail, MLP offload, calibrated 1593 threshold:** all five probes now return 48 correct tokens (`finish_reason=length`) where they returned 0 with `stop` before; and on the one probe that has a valid fresh reference (the first request of the earlier no-restore arm; its later probes were themselves cache restores and therefore broken), the restored output is byte-identical to the fresh padded output. So on this machine the restored path now matches the non-restored path.

**The two-bank non-determinism I reported above is the same bug.** With MLP 64 + GDN 48 (`tail_padding_min_tokens: 0`, so no padded tile involved), the natural partial-restore flow that gave 0/5 and 1/5 agreement between fresh processes now gives 5/5, and both runs are byte-identical to a fresh full prefill (5/5). That fits your mechanism: the cross-stream `_rope_deltas` dependency rides on every chunk of the restored path, so any ANE dispatch that host-waits on the producer buffer can read unwritten input — the padded tail hits it deterministically, and the co-scheduled banks hit it as a timing race. Per-machine byte-stable garbage on one path, run-to-run variance on the other, one cause.

That retires the "co-scheduled banks" reading from my earlier comment and the `gdn: false` suggestion with it; the correct mitigation before the fix lands is `tail_padding_min_tokens: 0`, and even that only removes the deterministic half.

Nice find. If it helps the PR, the arms above are two fresh processes each with the ANE state verified before every probe and `OMLX_SDPA256_TILED=0` pinned, on NAX hardware jundot does not have.

**Comment by daveycodez** [comment]
Validated hesara's root cause on a third rig, and in the configuration not yet covered above: M3 Ultra 512GB, Qwen3.8-27B-oQ4e-fp16-mtp, **Lightning MTP enabled**, dual-ANE MLP+GDN offload with CPU share (tuned split), tail_padding_min_tokens=1618, concurrent requests joining mid-turn, prefix-cache reuse against 13-16K cached contexts with odd-length uncached tails. Applied exactly one change to 0.6.3: mx.eval(lm._rope_deltas) immediately after the seed in _seed_text_only_mrope_delta_for_cached_prefill. Result: 5 rounds x 4 turns of the agentic repro (needle-verified retrieval each turn, sidecar join every turn) = 20/20 correct, zero corruption/ValueError/log anomalies. The same harness pre-fix produced the silent init-only empty completions reported earlier in this thread. So the lazy cross-stream _rope_deltas seed explains the MTP-enabled failure mode here as well — my earlier A/B that 'exonerated' ANE simply didn't hit the restore+tail trigger reliably (1-in-32). Great find; happy to re-run any additional probes on this rig if useful.

**Comment by jundot** [comment]
@daveycodez @hesara @beaglemoo Thanks for tracing this through the restore path and confirming it across different hardware. I merged #3279, which materializes the restore-only mRoPE seed before suffix prefill so the ANE path no longer inherits a lazy cross-stream dependency. The restored-prefix repros now complete correctly on the reported M3 Max, M5 Pro/NAX, and M3 Ultra setups, including the Lightning MTP case. This is fixed on main; please open a new report if the original corruption or empty-completion symptom returns.
